### ISSUE TITLE
Wrong query using EntityLoader manually with "where" param 

### ISSUE THREAD

**Issue by localnet** [mikro-orm/mikro-orm#273]
**Describe the bug**
I don't sure if I will get that I want with this code, but the generated query seems wrong (see the where clause):

```sql
[query] select `e0`.*, `e1`.`artist_id`, `e1`.`style_id` from `artist` as `e0` left join `artist_to_style` as `e1` on `e0`.`id` = `e1`.`artist_id` where 0 = 4 and `e1`.`style_id` in (2) [took 2 ms]
``` 

**To Reproduce**

```typescript
@Entity({ collection: "artist" })
export class ArtistEntity implements IdEntity<ArtistEntity> {
    @PrimaryKey()
    public id!: number;

    @ManyToMany(() => StyleEntity)
    public styles: Collection<StyleEntity>;

    public constructor() {
        this.styles = new Collection(this);
    }
}
```

```typescript
@Entity({ collection: "style" })
export class StyleEntity implements IdEntity<StyleEntity> {
    @PrimaryKey()
    public id!: number;

    @ManyToMany(() => ArtistEntity, "styles")
    public artists: Collection<ArtistEntity>;

    public constructor() {
        this.artists = new Collection(this);
    }
}
```

```typescript
@singleton()
export default class StyleStore {
    private styleRepository: EntityRepository<StyleEntity>;
    private entityLoader: EntityLoader;

    public constructor(@inject(OrmClient) ormClient: OrmClient) {
        this.styleRepository = ormClient.em.getRepository(StyleEntity);
        this.entityLoader = new EntityLoader(ormClient.em);
    }

    public async findArtists(ids: number[]): Promise<StyleEntity[][]> {
        const styles = await this.styleRepository.find({ artists: ids });

        await this.entityLoader.populate("StyleEntity", styles, ["artists"], { artists: ids });

        return[];
    }
}
```

**Expected behavior**
A where clause without:

```sql
where 0 = 4
```

**Additional context**
I am trying other possibilities to get more out of the ORM and get the same that with this:

```typescript
    public async findArtists(ids: number[]): Promise<StyleEntity[][]> {
        const results = await this.styleRepository.createQueryBuilder()
            .select("*").where({ artists: ids }).execute<any[]>();

        return ids.map((id) => results.filter((result) => result.artist_id === id)
            .map((entity) => this.styleRepository.map(entity)));
    }
```

**Versions**

| Dependency | Version |
| - | - |
| node | v12.13.1 |
| typescript | 3.7.2 |
| mikro-orm | 3.0.0-rc.3 |
| your-driver | mariadb |

PS:
The [documentation](https://mikro-orm.io/docs/nested-populate#using-entityloader-manually) seems have a typo error too:

```typescript
await loader.populate(Author, authors, ['books.tags']);
```

should be:

```typescript
await loader.populate('Author', authors, ['books.tags']);
```

**Comment by B4nan** [comment]
Why do you use `EntityLoader` manually?

    const styles = await this.styleRepository.find({ artists: ids }, ['artists']);

This should be the same I think.

**Comment by B4nan** [comment]
Also please try this:

    await this.entityLoader.populate("StyleEntity", styles, ["artists"], { artists: { $in: ids } });

Not 100% sure but I think entity loader expects the condition to be normalized to this form (which happens automatically if you use EM/QB API).

**Comment by localnet** [comment]
Yeah, that was. But yes, I can see it that this:

```typescript
const styles = await this.styleRepository.find({ artists: ids }, ["artists"]);
```

It is the same as this:

```typescript
const styles = await this.styleRepository.find({ artists: ids });
await this.entityLoader.populate("StyleEntity", styles, ["artists"], { artists: { $in: ids } });
```

I get all the artists of a style, when I only need a few. I am using GraphQl with [Dataloader](https://github.com/graphql/dataloader) for batching, but it require:

> There are a few constraints this function must uphold:
>   - The Array of values must be the same length as the Array of keys.
>   - Each index in the Array of values must correspond to the same index in the Array of keys.

For example with this GraphQL query:
```graphql
query {
  artists(ids: [2, 4]) {
    id
    about
    styles {
      id
      name
    }
  }
}
```

This schema:
```graphql
type Query {
    artists(ids: [ID]!): [Artist]!
}

type Artist {
    id: ID!
    about: String
    styles: [Style]!
}

type Style {
    id: ID!
    name: String!
}
```

And this resolver:
```typescript
export default {
    Query: {
        artists: async (...[, { ids }, { dataSources }]) => {
            return dataSources.artistProvider.getIds(ids.map(Number));
        },
    },
    Artist: {
        styles: async (...[{ id }, , { dataSources }]) => {
            return dataSources.styleProvider.getArtist(id);
        },
    }
};
```

First I get the "artists":
```typescript
export default class ArtistProvider extends DataSource {
    private artistStore: ArtistStore;
    private idLoader: DataLoader<number, ArtistEntity | undefined>;

    public constructor(@inject(ArtistStore) artistStore?: ArtistStore) {
        super();

        this.artistStore = artistStore as ArtistStore;
        this.idLoader = new DataLoader((ids) => this.artistStore.findIds(ids as number[]));
    }

    public async getIds(ids: number[]): Promise<Array<ArtistEntity | Error | undefined>> {
        return this.idLoader.loadMany(ids);
    }
}
```

```typescript
export default class ArtistStore {
    private artistRepository: EntityRepository<ArtistEntity>;

    public constructor(@inject(OrmClient) ormClient: OrmClient) {
        this.artistRepository = ormClient.em.getRepository(ArtistEntity);
    }

    public async findIds(ids: number[]): Promise<Array<ArtistEntity | undefined>> {
        const artists = await this.artistRepository.find(ids);

        return ids.map((id) => artists.find((artist) => artist.id === id));
    }
}
```

But now I need the "styles" of the "artists" in the same order, so I think my best is:

```typescript
export default class StyleProvider extends DataSource {
    private styleStore: StyleStore;
    private artistLoader: DataLoader<number, StyleEntity[]>;

    public constructor(@inject(StyleStore) styleStore?: StyleStore) {
        super();

        this.styleStore = styleStore as StyleStore;
        this.artistLoader = new DataLoader((ids) => this.styleStore.findArtists(ids as number[]));
    }

    public async getArtist(id: number): Promise<StyleEntity[]> {
        return this.artistLoader.load(id);
    }
}
```

```typescript
export default class StyleStore {
    private styleRepository: EntityRepository<StyleEntity>;

    public constructor(@inject(OrmClient) ormClient: OrmClient) {
        this.styleRepository = ormClient.em.getRepository(StyleEntity);
    }

    public async findArtists(ids: number[]): Promise<StyleEntity[][]> {
        const results = await this.styleRepository.createQueryBuilder()
            .select("*").where({ artists: ids }).execute<any[]>();

        return ids.map((id) => results.filter((result) => result.artist_id === id)
            .map((entity) => this.styleRepository.map(entity)));
    }
}
```

I have extended too much and really the issue is solved. Feel free to close it.

**Comment by B4nan** [comment]
I am a bit lost in that code as I have no GQL experience so far, so maybe I got it wrong. Is your problem that you load styles based by artist ids, and then want to populate their artists collections, but not fully (basically you want only those artists matching ids used to find the styles)? 

`EntityLoader` is not really for that use case, it populates the whole collections. 

Btw have you seen the other issue about dataloader #266? Maybe you will find that useful.

**Comment by B4nan** [comment]
Closing this now, but feel to continue with discussion.
